Add an agent injecting mutating webhook to the traffic manager.

thallgren opened the pull request:
## Description

This PR adds a mutating webhook that injects a traffic-agent to
pods that have the following annotation:

    telepresence.getambassador.io/traffic-agent: enabled

The installer will henceforth ignore deployments that have this
annotation and consider them to already have an agent installed.

The webhook is using `kates.List()` to list Service objects which
means that the cluster must be configured to allow the traffic-manager
to do that. A sample `k8s/agent-injector-rbac.yaml` is included that
provides the needed credentials.

A lot of code used by the agent installer is reused in the injector
and was therefore moved into the `agentinstall` package that they
both now share.
## Checklist

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes.
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

Commit:
Move Env to managerutil package and store it in the Context

vLuke:
 - Remove erroneous import from manager.go
 - Move envconfig_test.go into managerutil
 - Get envconfig_test.go to pass
 - Squash bugfix from a later commit
 - Tidy imports

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

Commit:
Move pkg/client/connector/installutil_*.go to pkg/install/*.go

And the necessary changes that go along with that.

vLuke:
 - Split in to separate commit

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

Commit:
Tidy imports

vLuke:
 - Split in to separate commit

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

Commit:
connector: Move some of the 'install' code in to separate files

This is in preparation to move them to a separate package so that they
can be shared with the manager.  This commit makes essentially no
modifications, and is a simple cut/paste job.  The one exception to
that is when pulling part of a function out to be its own function.

vLuke:
 - Split in to separate commit

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

Commit:
Add a program that will create the pems needed for agent-injection.

The program creates the crt.pem, key.pem, and ca.pem needed when
setting up the mutator webhook for agent auto injection.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

Commit:
Agent injection config and a service/deployment manifests for reference

Those manifests are not normally used but might be good to keep around
as a reference.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>
Signed-off-by: Thomas Hallgren <thomas@datawire.io>

donnyyung commented on k8s/agent-injector-rbac.yaml:
```diff
@@ -0,0 +1,27 @@
+# This should be the minium RBAC that is needed for users (not operators) to
```
I think this rbac is actually something that is needed for the operator who will be deploying the traffic-manager and not users (who won't need this role attached to their account and shouldn't need these permissions).

thallgren replied:
Doh, that comment is off. I'll change it.

thallgren replied:
Hmm, it's already changed. You need to refresh your browser :-)

LukeShu commented:
If `envconfig.go` is moving from `manager/` to `manager/managerutil/`, then `envconfig_test.go` should probably move too?

donnyyung commented:
I'm reliably getting: ```telepresence: error: Intercept "3b3c482d-4de9-4145-a343-f7d8fff0e3aa:echo-easy": Failed to establish intercept: intercept in error state AGENT_ERROR: no valid license & cannot connect to Ambassador Cloud for license check``` when using this branch with the Smart Agent.

Looking at the config, 
```
time="2021-06-10 15:25:43" level=info msg="config = {Config:{Name:echo-easy Namespace:default PodIP:10.42.0.45 AgentPort:9900 AppMounts:/tel_app_mounts AppPort:8080 ManagerHost:traffic-manager.ambassador ManagerPort:8081} EnvoyLogLevel:info EnvoyBaseID:1 UseHTTP2:true TerminatingTLSSecret: OriginatingTLSSecret: License: LicenseHost: LicensePublicKeys:{KeyMap:map[auth.datawire.io:-----BEGIN PUBLIC KEY----------END PUBLIC KEY----- beta-auth.datawire.io:-----BEGIN PUBLIC KEY----------END PUBLIC KEY-----]} ClusterID: CanConnectCloud:false SystemAHost: SystemAPort:}" func=main.Main file="github.com/datawire/telepresence2/proprietary/cmd/app-sidecar/sidecar.go:178" PID=1
```

Looking at the config, SystemAHost and SystemAPort (which the agent talks to the manager to get), aren't being correctly populated. I'll keep looking but wanted to raise the issue here in case you might know what in the changes might have affected this.

-- Edit --
I added a comment with what I believe to be the cause of the issue.

donnyyung commented on cmd/traffic/cmd/manager/service.go:
```diff
@@ -47,11 +48,10 @@ func (wall) Now() time.Time {
 	ret := &Manager{
 		ctx:   ctx,
 		clock: wall{},
-		env:   env,
```
I think this is the issue with the smart agent bc this function uses the config to get the System A info: https://github.com/telepresenceio/telepresence/blob/release/v2/cmd/traffic/cmd/manager/service.go#L100-L102

LukeShu commented:
Why not straight-up use the existing install "actions" system?

LukeShu commented:
"Add an agent injecting mutating webhook to the traffic manager" is difficult to review... it should probably be multiple commits.

LukeShu commented on cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:
```diff
@@ -0,0 +1,234 @@
+	"strings"
+
+	admission "k8s.io/api/admission/v1"
+	v1 "k8s.io/api/core/v1"
```
IMO, importing things as `v1` is an anti-pattern.  Say what it's v1 of! i.e. `corev1`

LukeShu commented on pkg/install/container.go:
```diff
@@ -0,0 +1,147 @@
+
+const envPrefix = "TEL_APP_"
+
+func AgentContainer(name, imageName string, appContainer *v1.Container, port v1.ContainerPort, appPort int, managerNs string) v1.Container {
```
That's a lot of args.  Maybe it should take a single struct?

At least wrap this to be 1 arg per line.

LukeShu commented on cmd/traffic/cmd/manager/managerutil/envconfig.go:
```diff
@@ -0,0 +1,42 @@
+	User        string `env:"USER,default="`
+	AgentImage  string `env:"TELEPRESENCE_AGENT_IMAGE,default="`
+	AgentPort   int32  `env:"TELEPRESENCE_AGENT_PORT,default=9900"`
+	Namespace   string `env:"MANAGER_NAMESPACE,default="`
```
I don't like having short envconfig field names like this.  If we thought it was worth it to put `MANAGER_` in the env-var name, then surely it's worth it to put `Manager` in the field name.

LukeShu commented on cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:
```diff
@@ -0,0 +1,234 @@
+			return strconv.Itoa(int(svcPort.Port))
+		}(), refPodName)
+
+	agentImage := env.AgentImage
+	if agentImage == "" {
+		agentImage = env.Registry + "/tel2:" + strings.TrimPrefix(version.Version, "v")
+	}
```
Consider doing this in `LoadEnv`

LukeShu commented on build/admission_controller_tls.go:
```diff
@@ -0,0 +1,71 @@
+package main
```
So... when does this need to get run?  Because nothing's running it right now.

thallgren replied:
This utility was created before I automated all of it in the installer. I thought it worth keeping around because it might help if we need to generate certs from the helm-chart.

LukeShu commented on cmd/traffic/cmd/manager/systema.go:
```diff
@@ -8,6 +8,8 @@ import (
 	"net"
 	"sync"
 
+	"github.com/telepresenceio/telepresence/v2/cmd/traffic/cmd/manager/managerutil"
+
 	"google.golang.org/grpc"
```
This should go further down, with the rest of the other telepresenceio imports.

thallgren replied:
So why does the linter _sometimes_ complain about this? It would be great if it captured all.

LukeShu commented on cmd/traffic/cmd/manager/service.go:
```diff
@@ -8,6 +8,8 @@ import (
 	"sort"
 	"time"
 
+	"github.com/telepresenceio/telepresence/v2/cmd/traffic/cmd/manager/managerutil"
+
 	"github.com/google/uuid"
```
This should go further down, with the rest of the telepresence.io imports.

LukeShu commented:
I want to say: I'm making a lot of style-related comments.  The point of them isn't to nitpick the style; I'm commenting when it's something that makes review harder and is costing me time.  For instance, I wanted to know what changed between `install_actions.go` and `install/container.go`, but that turned out to be needlessly difficult because `corev1` was renamed to `v1` and a bunch of brackets changed.  These things are style nitpicks that  on their own wouldn't be worth mentioning.  But they make the diff be 90% noise, so it's difficult to see what's actually changing.

LukeShu commented on pkg/client/config.go:
```diff
@@ -338,7 +338,7 @@ var defaultConfig = Config{
 		PrivateProxyDial:             5 * time.Second,
 		PrivateTrafficManagerAPI:     5 * time.Second,
-		PrivateTrafficManagerConnect: 20 * time.Second,
+		PrivateTrafficManagerConnect: 60 * time.Second,
```
This should be called out as a "change" in the changelog (with a note about why it changed).

LukeShu commented on pkg/client/envconfig.go:
```diff
@@ -19,8 +19,9 @@ type Env struct {
 
-	Registry   string `env:"TELEPRESENCE_REGISTRY,default=docker.io/datawire"`
-	AgentImage string `env:"TELEPRESENCE_AGENT_IMAGE,default="`
+	ManagerNamespace string `env:"TELEPRESENCE_MANAGER_NAMESPACE,default=ambassador"`
```
Supporting this is probably worth mentioning in the changelog.

thallgren replied:
Perhaps, but it isn't a change really. We've always supported that env using an `os.GetEnv()` in the installer.

thallgren replied:
I removed this change as it's not relevant to this PR anyway.

LukeShu reviewed (changes requested):
Please make this easier to review.  I spent all morning on it, and didn't get through it.

thallgren commented:
@LukeShu there are two reasons why I didn't use the installer "actions":

1. There's no need to `do`/`undo`. This is always just `do`
2. This is all about json patches. The actions actually modify manifests.

One could argue that the actions should be rewritten as json patches, but that didn't feel like something that needed to be done in this PR. Perhaps we should do that when we move the agent installer to the traffic-manager.

Commit:
Add a package "resource" with generic CRUD functions on kates.Object

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move agent service and port discovery over to the install package.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Increase the trafficManagerConnect timeout to 60 seconds.

Because this timeout often fails CI-tests and that situation will not
improve as we add more resources.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that test use default user when doing uninstall --everything

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move install, update, and delete of traffic-manager to resource package

Moves the code that previously installed, updated, and deleted the
traffic-manager from the connector package to the new resource
package. Much of this is copy/paste but since this is a more generic
mechanism, some parts were rewritten.

One notable change is that the license secret is now always mounted
but optional which should vouch for more consistent behavior when
that secret arrives or is removed (we didn't see that as a reason to
update the traffic-manager, but we should have since it actually added
or removed volumes from the deployment).

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add a mutator webhook service to the traffic-manager.

Adds an https service to the traffic-manager that is configured to
handle mutator webhooks. The service expects to find a TLS crt.pem
and key.pem in the /var/run/secrets/tls directory. The service
will only start if they are present (regular traffic-manager still
works if they aren't).

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add the agent-injector to the mutator webhook service.

Adds an agent injector which triggers on the annotation
"telepresence.getambassador.io/inject-traffic-agent".

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Accomodate for some changes in behavior when using injected agents.

1. An injected agent does not belong to a workload that has an action
annotation that explains how an installation is undone. That's OK, it
because there's nothing to uninstall in that workload.
2. The injected agent's workload must be considered up-to-date when
checking the agent.
3. The set of volumes for a pod that will be injected will already
contain the `/var/run/secrets/kubernetes.io/serviceaccount` volume so
appending the `/tel_app_mounts` to that volume must be avoided.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add automatic CRUD of manifests needed for agent-injection.

Adds CRUD operations for:
- traffic-manager ServiceAccount
- traffic-manager ClusterRole
- traffic-manager ClusterRoleBinding
- mutator-webhook Secret
- agent-injector Service
- agent-injector MutatingWebhookConfiguration,

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't use hardcoded "ambassador" namespace for generated TLS keys

These keys must be generated based on what namespace that the
agent-injector service is installed in.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add test of auto-injected traffic-agent.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

thallgren commented on pkg/install/resource/tm_cluster_role_binding.go:
```diff
@@ -0,0 +1,56 @@
+		{
+			Kind:      "ServiceAccount",
+			Name:      "traffic-manager",
+			Namespace: "ambassador",
```
This should not be hardcoded to "ambassador". A fix will arrive in subsequent commits.

thallgren commented on pkg/install/tls.go:
```diff
@@ -0,0 +1,94 @@
+		return nil, nil, nil, err
+	}
+
+	dnsNames := []string{"agent-injector", "agent-injector.ambassador", "agent-injector.ambassador.svc"}
```
The hardcoded "ambassador" will be replaced in a subsequent commit.

LukeShu commented on pkg/install/resource/instance.go:
```diff
@@ -0,0 +1,138 @@
+	"github.com/telepresenceio/telepresence/v2/pkg/client"
+)
+
+// An Instance exposes CRUD operations for a k8s resource such as
+// a Service.
+type Instance interface {
+	Create(context.Context) error
+	Exists(context.Context) (bool, error)
+	Delete(context.Context) error
+	Update(context.Context) error
```
CRUD, not CRDU! :P

LukeShu commented on pkg/install/resource/instance.go:
```diff
@@ -0,0 +1,138 @@
+	keyPem     []byte
+}
+
+// The scope is available through the context
+type scopeKey struct{}
+
+func withScope(ctx context.Context, scope *scope) context.Context {
+	return context.WithValue(ctx, &scopeKey{}, scope)
+}
+
+func getScope(ctx context.Context) *scope {
+	if sc, ok := ctx.Value(&scopeKey{}).(*scope); ok {
```
Those should just be `scopeKey{}`, not `&scopeKey{}`.

Honestly, I'm shocked it seems to work at all with the `&`.

thallgren replied:
Not intended. I'll remove that ampersand. IIRC, it works but the behavior is undefined in the spec.

LukeShu commented on CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 ### 2.3.1 (TBD)
 
+- Feature: Agents can now be installed using a mutator webhook
+- Change: The `trafficManagerConnect` timout default has changed from 20 seconds to 60 seconds, in order to facilitate TODO.
```
So I didn't actually finish writing that sentence.

thallgren replied:
I'll replace the `TODO` with `the extended time it takes to apply everything needed for the mutator webhook.`

LukeShu commented on pkg/install/resource/tm_deployment.go:
```diff
@@ -0,0 +1,183 @@
+
+const managerLicenseName = "systema-license"
+
+type tmDeployment struct {
+	found *kates.Deployment
+}
+
+var TrafficManagerDeployment Instance = &tmDeployment{}
```
So this is a global variable with state.  Which isn't a blocker, but... globals are a smell.

LukeShu replied:
Maybe we avoid that by having `GetTrafficManagerResources()` say `&tmDeployment{}` instead of `TrafficManagerDeployment`?

thallgren replied:
True that. Bad design. I'll fix it. There's no need for this one to be a variable.

LukeShu commented on pkg/install/resource/tm_namespace.go:
```diff
@@ -0,0 +1,44 @@
+
+type nsResource int
+
+var TrafficManagerNamespace Instance = nsResource(0)
+var TrafficManagerNamespaceKeep Instance = nsResource(1)
```
Can we change these from `var` to `const`?

thallgren replied:
No, because they are of type `Instance` which is an interface. We could of course cover that with an anonymous var instead.

thallgren replied:
I'll replace with:
```
const TrafficManagerNamespace = nsResource(0)
const TrafficManagerNamespaceKeep = nsResource(1)

var _ Instance = TrafficManagerNamespace
```
and do the same thing for the other integer base types.

LukeShu commented on pkg/install/resource/tm_svc.go:
```diff
@@ -0,0 +1,59 @@
+
+type tmSvc int
+
+var TrafficManagerSvc Instance = tmSvc(0)
```
Likewise, can this be `const`?

LukeShu commented on cmd/traffic/cmd/manager/internal/mutator/service.go:
```diff
@@ -0,0 +1,182 @@
+		w.WriteHeader(http.StatusOK)
+	})
+
+	server := &http.Server{
+		Addr:     ":" + strconv.Itoa(install.MutatorWebhookPortHTTPS),
+		ErrorLog: dlog.StdLogger(ctx, dlog.LogLevelError),
+		Handler:  mux}
+
+	dlog.Infof(ctx, "Mutating webhook service is listening on %v", server.Addr)
+	err := dutil.ListenAndServeHTTPSWithContext(ctx, server, certPath, keyPath)
```
If you use `dhttp` instead of the old deprecated `dutil`, then you don't have to configure `ErrorLog`.

LukeShu commented on cmd/traffic/cmd/manager/internal/mutator/service.go:
```diff
@@ -0,0 +1,182 @@
+	for _, skippedNs := range []string{
+		metav1.NamespacePublic,
+		metav1.NamespaceSystem,
+		"kube-node-lease",
```
You can type `"kube-node-lease"` as `corev1.NamespaceNodeLease` if you want.

LukeShu commented on pkg/install/resource/tm_deployment.go:
```diff
@@ -0,0 +1,183 @@
+			ReadOnly:  true,
+			MountPath: "/home/telepresence/",
+		},
+		{
+			Name:      "tls",
+			ReadOnly:  true,
+			MountPath: "/var/run/secrets/tls",
+		},
```
You added this in "Move install, update, and delete of traffic-manager to resource package" but probably didn't mean to add it until "Add a mutator webhook service to the traffic-manager".  You don't need to change it, but I got thrown through a loop for a minute.

LukeShu commented on cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:
```diff
@@ -0,0 +1,228 @@
+		// in which case we should use the AdmissionRequest namespace.
+		podNamespace = req.Namespace
+	}
+	podName := pod.Name
+	if podName == "" {
+		// It is very probable the pod was not yet assigned a name,
+		// in which case we should use the metadata generated name.
+		podName = pod.ObjectMeta.GenerateName
+	}
```
That... can't be right?  The Pod name is `GenerateName` plus a random suffix.
```
$ kubectl get pod -o yaml verylargejavaservice-6b696cb89f-blfdd |y2j|jq '{name: .metadata.name, generateName: .metadata.generateName}'
{
  "name":         "verylargejavaservice-6b696cb89f-blfdd",
  "generateName": "verylargejavaservice-6b696cb89f-"
}
```

thallgren replied:
Copy pasta from old impl. I'll remove it.

LukeShu commented on pkg/install/resource/injector_svc.go:
```diff
@@ -0,0 +1,58 @@
+				},
+			},
+		},
+	}
```
Is there a reason we don't just shove the port in to the existing manager service? Is there a reason to choose one approach over the other?

thallgren replied:
The webhook mechanism is hardcoded to port 443 and I didn't want to dedicate that port for this since it uses a generated TLS that is useless for other purposes. The traffic-manager service has `ClusterIP: None` so it's headless and cannot do proxying and I wanted a service that proxies 443 to  8443 in the manager. I'm sure there are other ways to do this. Let's revisit.

LukeShu reviewed (changes requested):
That was _so much_ easier to review, thank you!

I have 2 minor changes to request, and a question I want clarified.

- Change requested: the context key change
- Change requested: the changelog TODO
- Question: The Pod generateName thing.  Is it OK because the podName only gets used for user-messages?

Other than that, looks solid!  I mentioned a number of nitpicks and minor suggestions that you can act on or ignore.

LukeShu commented:
After you make the changes, feel free to merge without waiting for me to re-review.

Commit:
Changes in response to review

- Use corev1.NamespaceNodeLease instead of string literal
- Use dhttp instead of dutil
- Assume that pod.Name is always set
- Declare resource singletons as const instead of vars
- Don't use pointer to `scopeKey{}` when fetching context value
- Don't put stateful tmDeployment struct in var.
- Fix TODO comment in CHANGELOG.md

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

LukeShu reviewed (approved):
Hit it!

Commit:
Fix possible timing problem in auto-injected test.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Reinstate check for empty podName that defaults to generated name

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
